Add single-user authentication

O-Bezimo opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?             | no
| New feature?    | yes
| API breaks?       | no
| Deprecations?   | no
| Related tickets  | fixes #X, partially #Y, mentioned in #Z
| License         | Apache 2.0


### What's in this PR?
<!-- Explain the contents of the PR. Give an overview about the implementation, which decisions were made and why. -->

Our objective is to add the Single-User authentication in addition to the OIDC and LDAP authentication methods.

We added a SingleUserConfiguration object in the NifiCluster ressource.
The options are the following:
- Enabled boolean: set to true to add the according single-user configuration blocs in the nifi.properties template and the login_identity_provider template.
- SecretRef SecretReference: reference to a kubernetes secret containing the desired username and password.
- AuthorizerEnabled bool: set to true to use the single-user-authorizer. Set to false to use the default managed-authorizer.

We also added the reference to the SingleUserConfiguration in the helm folder.


### Why?
<!-- Which problem does the PR fix? (Please remove this section if you linked an issue above) -->

If neither OIDC or LDAP are configured, nifi forces a certificate authentication on all HTTP/HTTPS certificates.
We need another option so that we can have functionnal healthchecks on our loadbalancer generated by the ingress-controller.


### Additional context
<!-- Additional information we should know about (eg. edge cases, steps you followed to test the implementation) (Please remove this section if you don't need it) -->
We tested the code with an integration cluster hosted on gcp and telepresence as a proxy for our modified operator.

We also proofed some error cases:
- if SingleUserConfiguration.Enabled is false and SingleUserConfiguration.AuthorizerEnabled is true: don't configure the authorizer
- if SecretRef is empty: use auto generated username and password that would appear in the logs

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [x] Implementation tested
- [x] Error handling code meets the [guideline](docs/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)
- [x] Append changelog with changes

juldrixx commented on CHANGELOG.md:
```diff
@@ -10,11 +10,17 @@
 
 ### Removed
 
+## v1.3.0
+
+### Added
+
+- [PR #278](https://github.com/konpyutaika/nifikop/pull/278) - **[Operator]** Added the single-user-authentication method.
```
```suggestion
## Unreleased

### Added

- [PR #278](https://github.com/konpyutaika/nifikop/pull/278) - **[Operator]** Added the single-user-authentication method.

### Changed

### Fixed Bugs

### Deprecated

### Removed
```

The next release number is not defined right now.

juldrixx reviewed (commented):
Please delete the versionned documentation, the next version has not yet been decided.

And add the appropriate changes to the "unversioned" documentation to reflect your changes.

juldrixx commented on CHANGELOG.md:
```diff
@@ -10,11 +10,17 @@
 
 ### Removed
 
+## v1.3.0
+
+### Added
+
+- [PR #278](https://github.com/konpyutaika/nifikop/pull/278) - **[Operator]** Added the single-user-authentication method.
+
```
```suggestion
```suggestion
## Unreleased

### Added
- [PR #278](https://github.com/konpyutaika/nifikop/pull/278) - **[Operator]** Added the single-user-authentication method.

### Changed

### Fixed Bugs

### Deprecated

### Removed

```
```

O-Bezimo replied:
Modified accordingly

juldrixx commented:
@mh013370 Should secret keys be supplied in the operator configuration or forced as now?

O-Bezimo commented on CHANGELOG.md:
```diff
@@ -10,11 +10,17 @@
 ### Added
 
-- [PR #258](https://github.com/konpyutaika/nifikop/pull/141) - **[Helm Chart]** Upgraded helm-deployed HPA to v2 and added flowPosition to NiFiDataflow.
+- [PR #258](https://github.com/konpyutaika/nifikop/pull/141) - **[Operator]** Upgraded helm-deployed HPA to v2 and added flowPosition to NiFiDataflow.
```
Modified accordingly

O-Bezimo commented:
> @mh013370 Should secret keys be supplied in the operator configuration or forced as now?

Secret keys are supplied if singleUserConfiguration->enabled is true and if singleUserConfiguration->secretRef is not given in the nificluster.yml configuration. If enabled is true and secretRef is declared and exists in the cluster with a username and password, the operator will use these instead as credentials and add it in the login_identity_providers file.

mh013370 commented:
> @mh013370 Should secret keys be supplied in the operator configuration or forced as now?



> > @mh013370 Should secret keys be supplied in the operator configuration or forced as now?
> 
> Secret keys are supplied if singleUserConfiguration->enabled is true and if singleUserConfiguration->secretRef is not given in the nificluster.yml configuration. If enabled is true and secretRef is declared and exists in the cluster with a username and password, the operator will use these instead as credentials and add it in the login_identity_providers file.

If `SingleUserConfiguration.SecretRef` is optional, then it should be expressed as such in the spec by using a pointer (allowing you to check for `nil`) and adding the `// +optional` kubebuilder annotation.

mh013370 commented:
Why are there a whole new set of `unversioned` docs?

mh013370 commented on api/v1/nificluster_types.go:
```diff
@@ -100,11 +100,18 @@ type NifiClusterSpec struct {
 	// @TODO: Block Controller change
 }
 
+type SingleUserConfiguration struct {
```
These all need comments briefly explaining what they are and what they're used for. Especially the `SecretRef` & what its contents should look like.

O-Bezimo replied:
Done

mh013370 commented on api/v1/nificluster_types.go:
```diff
@@ -100,11 +100,18 @@ type NifiClusterSpec struct {
+type SingleUserConfiguration struct {
+	Enabled           bool            `json:"enabled,omitempty"`
+	AuthorizerEnabled bool            `json:"authorizerEnabled,omitempty"`
+	SecretRef         SecretReference `json:"secretRef,omitempty"`
```
Like mentioned in my other comment, recommend making this a pointer to a `SecretReference` and adding the `// +optional` kubebuilder annotation.

O-Bezimo replied:
Done

O-Bezimo commented:
> > @mh013370 Should secret keys be supplied in the operator configuration or forced as now?
> 
> > > @mh013370 Should secret keys be supplied in the operator configuration or forced as now?
> > 
> > 
> > Secret keys are supplied if singleUserConfiguration->enabled is true and if singleUserConfiguration->secretRef is not given in the nificluster.yml configuration. If enabled is true and secretRef is declared and exists in the cluster with a username and password, the operator will use these instead as credentials and add it in the login_identity_providers file.
> 
> If `SingleUserConfiguration.SecretRef` is optional, then it should be expressed as such in the spec by using a pointer (allowing you to check for `nil`) and adding the `// +optional` kubebuilder annotation.

Thank you for your answer, the modification have been made

O-Bezimo commented:
> > @mh013370 Should secret keys be supplied in the operator configuration or forced as now?
> 
> > > @mh013370 Should secret keys be supplied in the operator configuration or forced as now?
> > 
> > 
> > Secret keys are supplied if singleUserConfiguration->enabled is true and if singleUserConfiguration->secretRef is not given in the nificluster.yml configuration. If enabled is true and secretRef is declared and exists in the cluster with a username and password, the operator will use these instead as credentials and add it in the login_identity_providers file.
> 
> If `SingleUserConfiguration.SecretRef` is optional, then it should be expressed as such in the spec by using a pointer (allowing you to check for `nil`) and adding the `// +optional` kubebuilder annotation.

A pointer is added to Secret ref and the empty checks are made with nil as of now. An optional annotation is added and the same is made to a secretKeys struct allowing the final user to change the username and password keys. They will default to "username" and "password" if nil and the according information are added in the documentation and in the helm folder.

O-Bezimo commented:
> Why are there a whole new set of `unversioned` docs?

I switched the 1.3 versioned docs to unversioned which was a mistake. Since then I deleted it and moved the modifications in ./site/docs.

juldrixx commented on helm/nifi-cluster/values.yaml:
```diff
@@ -25,6 +25,19 @@ cluster:
+  singleUserConfiguration:
+    enabled: false
+    authorizerEnabled: false
+    secretRef: {
+      "name": "single-user-credentials",
+      "namespace": "nifi"
+    }
+    secretKeys: {
+      username: "username",
+      password: "password"
+    }
```
```suggestion
    secretRef:
      name: "single-user-credentials"
      namespace: "nifi"
    secretKeys:
      username: "username"
      password: "password"
```

O-Bezimo replied:
Modified

juldrixx replied:
It has not.

juldrixx commented on pkg/resources/templates/config/authorizers.go:
```diff
@@ -33,10 +41,19 @@ var AuthorizersTemplate = `{{- $nodeList := .NodeList }}
         <property name="Access Policy Provider">file-access-policy-provider</property>
     </authorizer>
 </authorizers>
+{{end}}
```
```suggestion
{{- end}}
```

O-Bezimo replied:
Modified

juldrixx commented on pkg/resources/templates/config/authorizers.go:
```diff
@@ -56,6 +73,7 @@ var EmptyAuthorizersTemplate = `<authorizers>
         <property name="Access Policy Provider">file-access-policy-provider</property>
     </authorizer>
 </authorizers>
+{{end}}
```
```suggestion
{{- end}}
```

O-Bezimo replied:
Modified

juldrixx commented on site/docs/5_references/1_nifi_cluster/1_nifi_cluster.md:
```diff
@@ -193,6 +194,15 @@ spec:
+| enabled           | boolean          | if set to true, we will enable single-user usage into nifi.properties configuration.                                                                     | No       | false   |
+| authorizerEnabled | boolean          | if set to true, we will use the single-user-authorizer instead of the managed-authorizer.                                                                | No       | false   |
+| secretRef         | [*SecretReference](../4_nifi_parameter_context#secretreference)   | the reference to the username and password. If not configured, will use auto generated username and password that will appear once in the logs at startup.      | No       | nil     |
+| secretKeys         | *SecretKeys   | the keys of the username and password.      | No       | {username= "username", password="password"}     |
```
A documentation section should be add for the SecretKeys doc to describe it. Like it's done for the SecretReference.
Directly in the 1_nifi_cluster.md file.

O-Bezimo replied:
Section added and link to the new section made in singleuserconfiguration

O-Bezimo commented on api/v1/nificluster_types.go:
```diff
@@ -101,10 +101,35 @@ type NifiClusterSpec struct {
+}
+
+type SecretKeys struct {
+	// +kubebuilder:default:=username
+	Username string `json:"username,omitempty"`
```
Modified

O-Bezimo commented on api/v1/nificluster_types.go:
```diff
@@ -101,10 +101,35 @@ type NifiClusterSpec struct {
+type SecretKeys struct {
+	// +kubebuilder:default:=username
+	Username string `json:"username,omitempty"`
+	// +kubebuilder:default:=password
+	Password string `json:"password,omitempty"`
```
Modified

O-Bezimo commented on helm/nifi-cluster/templates/nifi-cluster.yaml:
```diff
@@ -22,6 +22,19 @@ spec:
+      secretRef: {{ toYaml .Values.cluster.singleUserConfiguration.secretRef | nindent 6 }}
+    {{ end }}
+    {{ if .Values.cluster.singleUserConfiguration.secretKeys }}
+      secretRef: {{ toYaml .Values.cluster.singleUserConfiguration.secretKeys | nindent 6 }}
```
Modified

juldrixx commented on helm/nifi-cluster/values.yaml:
```diff
@@ -25,6 +25,19 @@ cluster:
+  singleUserConfiguration:
+    enabled: false
+    authorizerEnabled: false
+    secretRef: {
+      name: "single-user-credentials",
+      namespace: "nifi"
+    }
+    secretKeys: {
+      username: "username",
+      password: "password"
+    }
```
```suggestion
    secretRef:
      name: "single-user-credentials"
      namespace: "nifi"
    secretKeys:
      username: "username"
      password: "password"
```
